UART Driver/main.c

```diff
@@ -0,0 +1,307 @@
+static void MX_GPIO_Init(void);
+static void MX_USART2_UART_Init(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+//Creates the fijo and foji struct
+//The reason there are 2 different gpsCoordinates structs is to minimize the size of the fijo struct
+struct gpsCoordinatesFOJI{
+	double lattitude;
+	double longtitude;
+	double altitude;
+};
+
+struct gpsCoordinatesFIJO{
+	double lattitude;
+	double longtitude;
+};
+
+struct eulerAnglesOfPlane{
+	float yaw;
+	float pitch;
+	float roll;
+};
+
+//Should be 37 bytes but because of padding its really 48 bytes
+ struct foji{
+	uint8_t start;
+	struct gpsCoordinatesFOJI gpsCoord;
+	struct eulerAnglesOfPlane anglesOfPlane;
+};
+
+//Should be 20 bytes but because of padding its really 24 bytes
+struct fijo{
+	 uint8_t start;
+	_Bool takeoffCommand;
+	_Bool qrScanFlag;
+	_Bool detectFlag;
+	struct gpsCoordinatesFIJO gpsCoord;
+};
+
+const uint8_t STARTBYTE = 0x24;
+
+void sendFOJI(){
+	//Get out from telemetry manager
+	struct foji out;
+
+	HAL_UART_Transmit(&huart2, &out, sizeof(struct foji), HAL_MAX_DELAY);
+
+}
+
+void readFIJO(){
+	//Creates byte array
+	uint8_t buf[sizeof(struct fijo)];
+
+	//Initializes starting byte
+	buf[0] = STARTBYTE;
+
+	struct fijo in;
+	in.start = STARTBYTE;
+
+	//Receives a byte to check if its the same as STARTBYTE
+	uint8_t check;
+	HAL_UART_Receive(&huart2, &check, 1, HAL_MAX_DELAY);
+
+	if(check == STARTBYTE){
+		//Receive all the message bytes in buf
+		for(uint16_t i = 1; i < sizeof(struct fijo); i++)
+		{
+			HAL_UART_Receive(&huart2, &buf[i], 1, HAL_MAX_DELAY);
+		}
+
+		//Convert the byte array into an fijo struct
+		struct fijo *byteToStruct = (struct fijo*)buf;
+		in.takeoffCommand = byteToStruct->takeoffCommand;
+		in.qrScanFlag = byteToStruct->qrScanFlag;
+		in.detectFlag = byteToStruct->detectFlag;
+		in.gpsCoord.lattitude = byteToStruct->gpsCoord.lattitude;
+		in.gpsCoord.longtitude = byteToStruct->gpsCoord.longtitude;
+
+		//Send in into telemetry manager
+	}
+}
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  /* USER CODE BEGIN 2 */
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+    /* USER CODE END WHILE */
+
+    /* USER CODE BEGIN 3 */
+	  sendFOJI();
+	  readFIJO();
+  }
```

Comment:
As stated earlier, having this be an interrupt prevents this from blocking.

Is there a frequency of data transmission for this and is it handled elsewhere? This currently will send messages as fast as it can. You could use a timer to set the data frequency or use the telemetry manager's output rate to set it (Not sure how accurate that is). 
If the send function is called after each read function interrupt triggers than the output would be reasonably consistent.

Reply:
I think I would call the send function when required I just put it in the loop for sake of putting it in the loop. This would most likely be adjusted when it is implemented in telemetry manager.

Reply:
Makes sense, this way it gets called lots for debugging